Code/Framework/AzToolsFramework/AzToolsFramework/PaintBrush/PaintBrush.h

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/Memory/SystemAllocator.h>
+#include <AzToolsFramework/PaintBrush/PaintBrushNotificationBus.h>
+#include <AzToolsFramework/PaintBrushSettings/PaintBrushSettings.h>
+
+namespace AzToolsFramework
+{
+    //! PaintBrush contains the core logic for painting functionality.
+    //! It handles all of the specific calculations for determining the correct values to paint or smooth as a brush moves
+    //! across the world in a generic way. It relies on a listener to the PaintBrushNotificationBus to provide the other half
+    //! of the logic for fetching and applying the values themselves. 
+    //!
+    //! The general painting flow consists of the following:
+    //! 1. For each brush movement, the paintbrush sends OnPaint messages with the AABB of the region that has changed
+    //! and a callback for getting paint values for specific world positions.
+    //! 2. The listener calls the callback for each position in the region that it cares about.
+    //! 3. The paintbrush responds with the specific painted values for each of those positions based on the brush shape and settings.
+    //! This back-and-forth is needed so that we can keep a clean separation between the paintbrush and the listener. The paintbrush
+    //! doesn't have knowledge of which points in the world (or at which resolution) the listener needs, and the listener doesn't have
+    //! knowledge of the exact shape, size, and hardness of the paintbrush. The separation allows the paintbrush to add
+    //! features over time without having to change the listeners, and it enables more systems to use the paintbrush without requiring
+    //! specialized logic inside the paintbrush itself.
+    class PaintBrush
+    {
+
+    public:
+        AZ_RTTI(PaintBrush, "{D003175F-12BF-4E7E-BD51-2F4B010C9B5E}");
+        AZ_CLASS_ALLOCATOR(PaintBrush, AZ::SystemAllocator, 0);
+
+        PaintBrush() = delete;
+        PaintBrush(const PaintBrush&) = delete;
+        PaintBrush& operator=(const PaintBrush&) = delete;
+        explicit PaintBrush(const AZ::EntityComponentIdPair& entityComponentIdPair);
```

Comment:
question: Likely outside the scope of this PR but I wonder if it'd be possible to eliminate `entityComponentIdPair` from the core PaintBrush too so we could use it more easily in other viewports (e.g. The animation editor in future for painting vertex weights for example). We actually get around this with core manipulators by having an optional API to add an EntityComponentIdPair when needed which isn't required to be called. Not critical but could be something useful in future

Reply:
The entity component pair shouldn't prevent it from being used in other viewports, should it? It's just a mechanism for having a unique key for talking back and forth between the paintbrush and the component it's attached to. There are definitely other ways of generating a key and passing it around, but right now, since components are the only things consuming the paintbrush, the entity / component ID pair has been a pretty transparent way to connect the two. I'd prefer not to move to a more opaque key unless it's necessary.